File: binary_search.rb
def binary_search(array, target)
  left = 0
  right = array.size - 1

  while left + 1 < right
    mid = left + (right - left) / 2

    if array[mid] == target
      return mid
    elsif array[mid] > target
      right = mid
    else
      left = mid
    end
  end

  return left if array[left] == target
  return right if array[right] == target
end

File: binary_search.js
function binarySearch(array, target) {
  var left = 0;
  var right = array.length - 1;

  while (left + 1 < right) {
    var mid = Math.floor(left + (right - left) / 2);

    if (array[mid] == target) {
      return mid;
    } else if (array[mid] < target) {
      left = mid;
    } else {
      right = mid;
    }
  }

  if (array[left] == target) { return left; }
  if (array[right] == target) { return right; }
}

console.log(binarySearch([0, 2, 4, 5, 7, 8, 10], 7) === 4);
console.log(binarySearch([0, 2, 4, 5, 7, 8, 10], 0) === 0);
